### PR TITLE
Add max_mean aggregation to evaluation.

### PR DESCRIPTION
**Pull request by janfrancu**
- [x] initial draft

- [x] check if output is correct

- [x] ~~criteria on number of samples/repetition, i.e. implement `min_samples`~~ does not make much sense in max-mean aggr

- [x] ~~modify `evaluate_performance` script to take this new aggregation into account~~ exposed only the default mean-max method

### CODE REVIEW

File: src/evaluation/stats.jl [hunk 1]
```diff
@@ -170,3 +172,70 @@ function aggregate_stats(df::DataFrame, criterion_col=:val_auc;
 	end
 	vcat(results...)
 end
+
+
+"""
+aggregate_stats_max_mean(df::DataFrame, criterion_col=:val_auc; undersample=Dict("ocsvm" => 100))
```
Comment (vitskvara) [inline src/evaluation/stats.jl]: Change the docstring to downsample (in the original aggregate stats its missing as well).

File: src/evaluation/stats.jl [hunk 2]
```diff
@@ -170,3 +172,70 @@ function aggregate_stats(df::DataFrame, criterion_col=:val_auc;
...
+
+Chooses the best hyperparameters for each seed/anomaly_class combination by `criterion_col`
+and then aggregates the metrics over seed/anomaly_class to get the final results. The output 
+is a DataFrame of maximum #datasets*#models rows with
+columns of different types
+- identifiers - `dataset`, `modelname`, `phash`, `parameters`
```
Comment (vitskvara) [inline src/evaluation/stats.jl]: When i tried to run this, I did not get the columns with phash and parameters.

Comment (vitskvara) [reply src/evaluation/stats.jl]: Which makes sense, since they are now not universal for all selected models. But you should edit the docstring.

**Comment by vitskvara** [comment]
So I wrote the following test where I recompute the numbers in a different, slightly more transparent but uglier way. The numbers fit exactly for most models and are slightly off for "knn" or "vae", which I think is caused by more than one seed for theses two models. Correct me if I am wrong. From this, I would approve the pull request.

```julia
using DrWatson
@quickactivate
using CSV
using BSON
using Random
using FileIO
using DataFrames
using Statistics

using GenerativeAD.Evaluation: _prefix_symbol, PAT_METRICS, aggregate_stats
using GenerativeAD.Evaluation: rank_table, print_rank_table, aggregate_stats_max_mean

filename = "/home/skvarvit/generativead/GenerativeAD.jl/data/evaluation/images_eval.bson"
criterion_metric =  "val_auc"
downsample=Dict()
df = load(filename)[:df]
df_agg = aggregate_stats_max_mean(df, Symbol(criterion_metric), downsample=downsample)

# test if the numbers fit
dataset = "CIFAR10"
modelname = "knn"
subdf = filter(r->r[:dataset]==dataset && r[:modelname]==modelname, df)
smeasure = "val_auc"
pmeasure = "tst_auc"
res = []
for ac in 1:10
	for seed in 1:1
		ssdf = sort(filter(r->r[:seed] == seed && r[:anomaly_class]==ac, subdf), order(smeasure, rev=true))
		push!(res, ssdf[1, pmeasure])
	end
end
orig_df = filter(r->r[:dataset]==dataset && r[:modelname]==modelname, df_agg_mm)[!, vcat([50, 51], 1:49)]

# these two numbers should be the same
mean(res)
orig_df[pmeasure]
```

**Comment by vitskvara** [comment]
Actually by adding all 5 seeds and a check for empty dataframe I now conclude that the reported numbers are correct.

File: src/evaluation/stats.jl [hunk 3]
```diff
@@ -170,3 +172,70 @@ function aggregate_stats(df::DataFrame, criterion_col=:val_auc;
...
+agg_cols = vcat(_prefix_symbol.("val", BASE_METRICS), _prefix_symbol.("tst", BASE_METRICS))
+agg_cols = vcat(agg_cols, _prefix_symbol.("val", PAT_METRICS), _prefix_symbol.("tst", PAT_METRICS))
+top10_std_cols = _prefix_symbol.(agg_cols, "top_10_std")
+
+agg_keys = ("anomaly_class" in names(df)) ? [:seed, :anomaly_class] : [:seed]
+# choose best for each seed/anomaly_class cobination and average over them
+results = []
+for (dkey, dg) in pairs(groupby(df, :dataset))
+    for (mkey, mg) in pairs(groupby(dg, :modelname))
+        partial_results = []
```
Comment (vitskvara) [inline src/evaluation/stats.jl]: If we have agreed on seed 1 only, you should filter here somehow.

**Review by vitskvara — APPROVED** [review]

File: src/evaluation/stats.jl (hunk 3 above)
Comment (janfrancu) [reply src/evaluation/stats.jl]: I hesitate to put such filter into the aggregation itself and leave this to the user to filter beforehand.

File: src/evaluation/stats.jl (hunk 2 above)
Comment (janfrancu) [reply src/evaluation/stats.jl]: Updated.

**Comment by janfrancu** [comment]
Thanks for the feedback. Merging.

**Pull request merged**